components/dependency_injection/autowiring.rst

```diff
@@ -0,0 +1,411 @@
+.. index::
+    single: DependencyInjection; Autowiring
+
+Using the Autowiring
+====================
+
+Autowiring allows to register services in the container with minimal configuration.
+It is practical in the field of Rapid Application Development, when designing prototypes
+and in early stages of large projects. It makes it easy to bootstrap an app service
+graph and eases refactoring.
+
+Let’s see how it works. To do so we will build a fake API publishing statutes on a Twitter
+feed obfuscated with ROT13 (a special case of the Caesar cipher).
+
+Start by creating a ROT13 transformer class::
+
+    // src/AppBundle/Rot13Transformer.php
+
+    namespace AppBundle;
+
+    class Rot13Transformer
+    {
+        public function transform($value)
+        {
+            return str_rot13($value);
+        }
+    }
+
+And now a Twitter client using this transformer::
+
+    // src/AppBundle/TwitterClient.php
+
+    namespace AppBundle;
+
+    class TwitterClient
+    {
+        private $rot13Transformer;
+
+        public function __construct(Rot13Transformer $rot13Transformer)
+        {
+            $this->rot13Transformer = $rot13Transformer;
+        }
+
+        public function tweetInRot13($user, $key, $status)
+        {
+            $transformedStatus = $this->rot13Transformer->transform($status);
+
+            // Connect to Twitter and send the encoded status
+        }
+    }
+
+
+The Dependency Injection Component is now able to automatically register the dependencies
+of this ``TwitterClient`` class. The ``twitter_client`` service definition just
+need to be marked as autowired:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # app/config/services.yml
+
+        services:
+            twitter_client:
+                class:    AppBundle\TwitterClient
+                autowire: true
+
+    .. code-block:: xml
+
+        <!-- app/config/services.xml -->
+
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <service id="twitter_client" class="AppBundle\TwitterClient" autowire="true" />
+            </services>
+        </services>
+
+    .. code-block:: php
+
+        use Symfony\Component\DependencyInjection\Definition;
+
+        // ...
+        $definition = new Definition('AppBundle\TwitterClient');
+        $definition->setAutowired(true);
+
+        $container->setDefinition('twitter_client', $definition);
+
+The autowiring subsystem will parse the constructor of the ``TwitterClient`` class
```

Comment:
(L92 - 102) The autowiring subsystem will detect the dependencies of the ``TwitterClient`` class by parsing the constructor of the class. For instance it will find here the dependency to an instance of a ``Rot13Transformer``. If an existing service definition (and only one – see below) is of the required type, this service will be injected. If it not the case (like in this example), the subsystem is smart enough to automatically register a private service for the ``Rot13Transformer`` class and set it as first argument of the ``twitter_client`` service. Again, it can work only if there is one class of the given type. If there are several classes of the same type, you must use an explicit service definition or register a default implementation.